File: src/nacos_mcp_router/router.py
import asyncio
import json
import os

import anyio
from mcp import types
from mcp.client.stdio import get_default_environment
from mcp.server import Server
from mcp.server.fastmcp import FastMCP

from .logger import NacosMcpRouteLogger
from .mcp_manager import McpUpdater
from .nacos_http_client import NacosHttpClient
from .router_types import ChromaDb
from .router_types import CustomServer

nacos_addr = os.getenv("NACOS_ADDR","127.0.0.1:8848")
nacos_user_name = os.getenv("NACOS_USERNAME","nacos")
nacos_password = os.getenv("NACOS_PASSWORD","")
nacos_http_client = NacosHttpClient(nacosAddr=nacos_addr if nacos_addr != "" else "127.0.0.1:8848", userName=nacos_user_name if nacos_user_name != "" else "nacos",passwd=nacos_password)
chroma_db_service = ChromaDb()
mcp_updater = McpUpdater(nacosHttpClient=nacos_http_client, chromaDbService=chroma_db_service, update_interval=60)
mcp_servers_dict = {}

router_logger = NacosMcpRouteLogger.get_logger()

async def search_mcp_server(task_description: str, key_words: list[str]) -> str:
  """
    Name:
        search_mcp_server

    Description:
        执行任务前首先使用本工具。根据任务描述及关键字搜索mcp server，制定完成任务的步骤。

    Args:
        task_description (string): 用户任务描述，使用中文
        key_words (list): 字符串数组，用户任务关键字，使用中文,可以为多个，最多为2个
  """
  try:
    mcp_servers1 = []
    for key_word in key_words:
      mcps = mcp_updater.search_mcp_by_keyword(key_word)
      if len(mcps) > 0:
        for mcp in mcps:
          mcp_servers1.append(mcp)

    if len(mcp_servers1) < 5:
      key_words.append(task_description)
      mcp_servers2 = mcp_updater.getMcpServer(task_description,5-len(mcp_servers1))
      for mcp in mcp_servers2:
        mcp_servers1.append(mcp)

    result = {}
    for mcpServer in mcp_servers1:
      dct = {}
      dct['name'] = str(mcpServer.get_name())
      dct['description'] = str(mcpServer.get_description())
      result[str(mcpServer.get_name())] = dct

    content = json.dumps(result, ensure_ascii=False)

    jsonString = "## 获取" + task_description + "的步骤如下：\n" + '''
    ### 1. 当前可用的mcp server列表为：''' + content + '''
    \n ### 2. 从当前可用的mcp server列表中选择你需要的mcp server调add_mcp_server工具安装mcp server
    '''
    return jsonString
  except Exception as e:
    router_logger.warning("failed to search_mcp_server: " + task_description, exc_info=e)
    jsonString = "failed to search mcp server for " + task_description
    return jsonString


async def use_tool(mcp_server_name: str, mcp_tool_name: str, params:dict) -> str:
  try:
    if mcp_server_name not in mcp_servers_dict:
      router_logger.warning("mcp server {} not found, use search_mcp_server to get mcp servers".format(mcp_server_name))
      return "mcp server not found, use search_mcp_server to get mcp servers"

    mcp_server = mcp_servers_dict[mcp_server_name]
    if mcp_server.healthy():
      response = await mcp_server.execute_tool(mcp_tool_name, params)
    else:
      del mcp_servers_dict[mcp_server_name]
      return  "mcp server is not healthy, use search_mcp_server to get mcp servers"
    return str(response.content)
  except Exception as e:
    router_logger.warning("failed to use tool: " + mcp_tool_name, exc_info=e)
    return "failed to use tool: " + mcp_tool_name

async def add_mcp_server(mcp_server_name: str) -> str:
  """
  安装指定的mcp server
  :param mcp_server_name: mcp server名称
  :return: mcp server安装结果
  """
  try:
    mcp_server = nacos_http_client.get_mcp_server_by_name(mcp_server_name)
    if mcp_server is None or mcp_server.description == "":
      mcp_server = mcp_updater.get_mcp_server_by_name(mcp_server_name)

    if mcp_server is None:
      return mcp_server_name + " is not found" + ", use search_mcp_server to get mcp servers"

    disenabled_tools = {}
    tools_meta = mcp_server.mcp_config_detail.tool_spec.tools_meta
    for tool_name in tools_meta:
      meta = tools_meta[tool_name]
      if not meta.enabled:
        disenabled_tools[tool_name] = True

    if mcp_server_name not in mcp_servers_dict:
      env = get_default_environment()
      if mcp_server.agentConfig is None:
        mcp_server.agentConfig = {}
      if 'mcpServers' not in mcp_server.agentConfig or mcp_server.agentConfig['mcpServers'] is None:
        mcp_server.agentConfig['mcpServers'] = {}

      mcp_servers = mcp_server.agentConfig["mcpServers"]
      for key, value in mcp_servers.items():
        server_config = value
        if 'env' in server_config:
          for k in server_config['env']:
            env[k] = server_config['env'][k]
        server_config['env'] = env
        if 'headers' not in server_config:
          server_config['headers'] = {}

      server = CustomServer(name=mcp_server_name,config=mcp_server.agentConfig)
      await server.wait_for_initialization()
      if server.healthy():
        mcp_servers_dict[mcp_server_name] = server

    server = mcp_servers_dict[mcp_server_name]

    tools = await server.list_tools()
    tool_list = []
    for tool in tools:
      if tool.name in disenabled_tools:
        continue
      dct = {}
      dct['name'] = tool.name
      dct['description'] = tool.description
      dct['inputSchema'] = tool.inputSchema
      tool_list.append(dct)

    nacos_http_client.update_mcp_tools(mcp_server_name,tools)

    result = "1. " + mcp_server_name + "安装完成, tool 列表为: " + json.dumps(tool_list, ensure_ascii=False) +  "\n 2." + mcp_server_name + "的工具需要通过nacos-mcp-router的use_tool工具代理使用"
    return result
  except Exception as e:
    router_logger.warning("failed to install mcp server: " + mcp_server_name, exc_info=e)
    return "failed to install mcp server: " + mcp_server_name

def main() -> int:
  app = Server("nacos_mcp_router")

  @app.call_tool()
  async def call_tool(
          name: str, arguments: dict
  ) -> list[types.TextContent | types.ImageContent | types.EmbeddedResource]:
    match name:
      case "search_mcp_server":
        content = await search_mcp_server(arguments["task_description"], arguments["key_words"])
        return [types.TextContent(type="text", text=content)]
      case "add_mcp_server":
        content = await add_mcp_server(arguments["mcp_server_name"])
        return [types.TextContent(type="text", text=content)]
      case "use_tool":
        content = await use_tool(arguments["mcp_server_name"],arguments["mcp_tool_name"], arguments["params"])
        return [types.TextContent(type="text", text=content)]
      case _:
        return [types.TextContent(type="text", text="not implemented tool")]
    

  @app.list_tools()
  async def list_tools() -> list[types.Tool]:
    return [
      types.Tool(
        name="search_mcp_server",
        description="执行任务前首先使用本工具。根据任务描述及关键字搜索mcp server, 制定完成任务的步骤。",
        inputSchema={
          "type": "object",
          "required": ["task_description", "key_words"],
          "properties": {
            "task_description": {
              "type": "string",
              "description": "用户任务描述 ",
            },
            "key_words": {
              "type": "list[string]",
              "description": "用户任务关键字，可以为多个，最多为2个"
            }
          },
        },
      ),
      types.Tool(
        name="add_mcp_server",
        description="安装指定的mcp server",
        inputSchema={
          "type": "object",
          "required": ["mcp_server_name"],
          "properties": {
            "mcp_server_name": {
              "type": "string",
              "description": "MCP Server名称"
            }
          }
        }
      ),
      types.Tool(
        name="use_tool",
        description="使用某个MCP Server的工具",
        inputSchema={
           "type": "object",
          "required": ["mcp_server_name","mcp_tool_name","params"],
          "properties": {
            "mcp_server_name": {
              "type": "string",
              "description": "需要使用的MCP Server名称"
            },
            "mcp_tool_name":{
              "type": "string",
              "description": "需要使用的MCP Server工具名称"
            },
            "params": {
              "type": "map",
              "description": "需要使用的MCP Server工具的参数"
            }
          }
        }
      )
    ]

  from mcp.server.stdio import stdio_server

  async def arun():
    async with stdio_server() as streams:
      await app.run(
        streams[0], streams[1], app.create_initialization_options()
      )

  anyio.run(arun)
  
  return 0

File: src/nacos_mcp_router_typescript/src/router.ts
import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { StdioServerTransport } from "@modelcontextprotocol/sdk/server/stdio.js";
import { NacosHttpClient } from "./nacos_http_client";
import { McpManager } from "./mcp_manager";
import { logger } from "./logger";
import { z } from "zod";
import { ErrorCode, McpError } from "@modelcontextprotocol/sdk/types.js";
import { ChromaDb, NacosMcpServer } from "./router_types";
// import { Server } from "@modelcontextprotocol/sdk/server/index.js";
import { Transport } from "@modelcontextprotocol/sdk/shared/transport.js";

const MCP_SERVER_NAME = "nacos-mcp-router";

export interface RouterConfig {
  nacos: {
    serverAddr: string;
    username: string;
    password: string;
  };
  mcp: {
    host: string;
    port: number;
    authToken?: string;
  };
}

interface ServiceInfo {
  name: string;
  description: string;
}

export class Router {
  private nacosClient: NacosHttpClient;
  private mcpManager: McpManager;
  private chromaDb: ChromaDb;
  private config: RouterConfig;
  private serviceCache: Map<string, ServiceInfo> = new Map();
  private updateInterval: NodeJS.Timeout | null = null;
  private mcpServer: McpServer;

  constructor(config: RouterConfig) {
    this.config = config;
    const {serverAddr, username, password} = config.nacos;
    this.nacosClient = new NacosHttpClient(serverAddr, username, password);
    this.chromaDb = new ChromaDb();
    this.mcpManager = new McpManager(this.nacosClient, this.chromaDb, 60000);
    this.mcpServer = new McpServer({
      name: MCP_SERVER_NAME,
      version: "1.0.0",
    });
    this.registerMcpTools();
  }

  public async registerMcpTools() {
    try {
      this.mcpServer.tool(
        "SearchMcpServer",
        `根据任务描述及关键字搜索mcp server，制定完成任务的步骤;Args:task_description: 用户任务描述，使用中文;key_words: 字符串数组，用户任务关键字，使用中文,可以为多个，最多为2个`,
        { taskDescription: z.string(), keyWords: z.string().array().nonempty({
          message: "Can't be empty!",
        }).max(2) },
        async ({ taskDescription, keyWords }) => {
          try {
            const mcpServers1: NacosMcpServer[] = [];
            
            // 根据关键字搜索MCP服务器
            for (const keyWord of keyWords) {
              const mcps = await this.mcpManager.searchMcpByKeyword(keyWord);
              if (mcps.length > 0) {
                mcpServers1.push(...mcps);
              }
            }

            // 如果找到的服务器数量少于5个，使用任务描述进行从向量库补充搜索
            if (mcpServers1.length < 5) {
              const additionalServers = await this.mcpManager.getMcpServer(taskDescription, 5 - mcpServers1.length);
              mcpServers1.push(...additionalServers);
            }

            // 构建结果
            const result: Record<string, { name: string; description: string }> = {};
            for (const mcpServer of mcpServers1) {
              result[mcpServer.getName()] = {
                name: mcpServer.getName(),
                description: mcpServer.getDescription()
              };
            }

            const content = JSON.stringify(result, null, 2);
            const jsonString = `## 获取${taskDescription}的步骤如下：
### 1. 当前可用的mcp server列表为：
${content}
### 2. 从当前可用的mcp server列表中选择你需要的mcp server调AddMcpServer工具安装mcp server`;

            return {
              content: [{
                type: "text",
                text: jsonString
              }]
            };
          } catch (error) {
            logger.warn(`failed to search_mcp_server: ${taskDescription}`, error);
            return {
              content: [{
                type: "text",
                text: `failed to search mcp server for ${taskDescription}`
              }]
            };
          }
        }
      );
      this.mcpServer.tool(
        "UseTool",
        '使用指定MCP服务器上的工具。需要先通过AddMcpServer安装MCP服务器，然后才能使用其工具。',
        { mcpServerName: z.string(), toolName: z.string(), params: z.record(z.string(), z.any()) },
        async ({ mcpServerName, toolName, params }) => {
          try {
            const result = await this.mcpManager.useTool(mcpServerName, toolName, params);
            return {
              content: [{
                type: "text",
                text: JSON.stringify(result)
              }]
            };
          } catch (error) {
            logger.error(`Failed to use tool ${toolName} from server ${mcpServerName}:`, error);
            // throw new McpError(ErrorCode.InternalError, `Failed to use tool ${toolName} from server ${mcpServerName}`);
            return {
              content: [{
                type: "text",
                text: `Failed to use tool ${toolName} from server ${mcpServerName}`
              }]
            };
          }
        }
      );
      this.mcpServer.tool(
        "AddMcpServer",
        `安装指定的mcp server, return mcp server安装结果`,
        { mcpServerName: z.string() },
        async ({ mcpServerName }) => {
          try {
            const result = await this.mcpManager.addMcpServer(mcpServerName);
            return {
              content: [{
                type: "text",
                text: JSON.stringify(result)
              }]
            };
          } catch (error) {
            logger.error(`Failed to add mcp server ${mcpServerName}:`, error);
            throw new McpError(ErrorCode.InternalError, `Failed to add mcp server ${mcpServerName}`);
          }
        }
      );
    } catch (error) {
      logger.error("Failed to register MCP tools:", error);
      throw new McpError(ErrorCode.InternalError, "Failed to register MCP tools:", error);
    }
  }

  public async start(replaceTransport?: Transport) {
    try {
      const transport = new StdioServerTransport();
      if (replaceTransport) {
        this.mcpServer!.connect(replaceTransport);
      } else {
        await this.mcpServer!.connect(transport);
      }
    } catch (error) {
      logger.error("Failed to start Nacos MCP Router:", error);
      // throw error;
    }
  }
}
